### PR TITLE
Convert metrics.proto to proto3

### PR DESCRIPTION
**Pull request by juliusv**
I even got the Java stuff to build locally, but I think someone else will have to deploy / release that (@brian-brazil?).

### CODE REVIEW

**Comment by brian-brazil** [comment]
I don't think we ever released the histogram support for Java, so I wouldn't worry about it.

File: metrics.proto [hunk 1]
```diff
@@ -30,52 +30,52 @@ enum MetricType {
...
-  optional double value = 1;
+  double value = 1;
 }
 
 message Histogram {
-  optional uint64 sample_count = 1;
-  optional double sample_sum   = 2;
-  repeated Bucket bucket       = 3; // Ordered in increasing order of upper_bound, +Inf bucket is optional.
+  uint64   sample_count  = 1;
+  double   sample_sum    = 2;
```
Comment (beorn7) [inline metrics.proto]: This whitespace alignment is not consistent with the one elsewhere.

Either align all the _names_ (not types) everywhere, or don't other to align the repeated types. But this one aligns the names to the Bucket type below.

Comment (juliusv) [inline metrics.proto]: Ah right, done.

File: metrics.proto [hunk 2]
```diff
@@ -30,52 +30,52 @@ enum MetricType {
...
-  optional double value = 1;
+  double value = 1;
 }
 
 message Histogram {
-  optional uint64 sample_count = 1;
-  optional double sample_sum   = 2;
-  repeated Bucket bucket       = 3; // Ordered in increasing order of upper_bound, +Inf bucket is optional.
+  uint64 sample_count  = 1;
+  double sample_sum    = 2;
```
Comment (beorn7) [inline metrics.proto]: Now either outdent the index number by one in these two or indent by one (to match below). It looks like we are indenting the indices enough to align them all in general.

File: metrics.proto [hunk 3]
```diff
@@ -30,52 +30,52 @@ enum MetricType {
...
-  optional double value    = 2;
+  double quantile = 1;
+  double value    = 2;
 }
 
 message Summary {
-  optional uint64   sample_count = 1;
-  optional double   sample_sum   = 2;
+  uint64 sample_count = 1;
+  double sample_sum   = 2;
```
Comment (beorn7) [inline metrics.proto]: Make indices align with line 48

File: metrics.proto (hunk 2 above)
Comment (juliusv) [inline metrics.proto]: Done.

File: metrics.proto (hunk 3 above)
Comment (juliusv) [inline metrics.proto]: Done.

**Comment by beorn7** [comment]
:+1: 
My apology for all the nits. ;)

**Comment by juliusv** [comment]
Holding off on the merge until the discussion in https://github.com/prometheus/client_model/issues/13 is settled

**Comment by fsouza** [comment]
Now that prometheus/common#52 is merged, can we get this one in? :D

**Comment by juliusv** [comment]
@fsouza No, if the discussion in #13 results in the conclusion that proto3 introduces a breaking change and general regression, we might have to cancel the move to proto3.

**Comment by fsouza** [comment]
@juliusv then prometheus/common#52 should be reverted. Master is broken there.

**Comment by beorn7** [comment]
I guess with the current role of protobuf in the Prometheus ecosystem, there is really no need to think about proto3 migration, especially if that needs changes in the detailed protobuf spec, where the latter is doubly irrelevant as (1) the Prometheus server is currently not supporting protobuf and (2) OpenMetrics will present a different protobuf spec.

Once OpenMetrics with its protobuf spec is finalized, we can revisit the issue. (For example, if we go all in with OpenMetrics, it's proto representation will be part of the standard and there would no need to have it in a Prometheus-specific repo at all. Code generation could then be done by each using repo in the way they prefer, see how it is handled for remote write in prometheus/prometheus.)

Closing this super stale PR in any case.

**Pull request closed without merging**